### PR TITLE
CLIMB standard defaults

### PR DESCRIPTION
**Pull request by smillst**
The nullness tests are failing because of the change in defaults for the Rawness and Initialization type systems.  The default for implicit upper bounds use to be @NonRaw and @Initialized, but now it is @Raw and @UnknownInitialization.  The fake enum and OIGJ  tests are failing for similar reasons. 

Should we update the tests (and the annotated JDK) to account for these defaulting changes? Alternatively, we could change the default of explicit upper bounds to top.

### CODE REVIEW

**Comment by wmdietl** [comment]
I don't think it makes sense to use top as the default for implicit upper bounds in the Rawness & Initialization Checkers.
I would change the default for implicit upper bounds for these type systems and use NonRaw and Initialized, respectively. I would not change the default for explicit upper bounds.

File: checker/src/org/checkerframework/checker/fenum/FenumAnnotatedTypeFactory.java [hunk 1]
```diff
@@ -34,6 +35,15 @@ public FenumAnnotatedTypeFactory(BaseTypeChecker checker) {
         this.postInit();
     }
 
+    @Override
+    protected void addCheckedCodeDefaults(QualifierDefaults defs) {
+        // An error is issued if the type of a thrown exception
+        // is not @FenumUnqualified.
+        defs.addCheckedCodeDefault(FENUM_UNQUALIFIED, DefaultLocation.EXCEPTION_PARAMETER);
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/fenum/FenumAnnotatedTypeFactory.java]: Why is this here instead of using the declarative features on FenumUnqualified?

File: checker/src/org/checkerframework/checker/igj/IGJAnnotatedTypeFactory.java [hunk 2]
```diff
@@ -180,6 +182,14 @@ public IGJAnnotatedTypeFactory(BaseTypeChecker checker) {
     }
 
     @Override
+    protected void addCheckedCodeDefaults(QualifierDefaults defs) {
+        // An error is issued if the type of a thrown exception
+        // is not @Mutable
+        defs.addCheckedCodeDefault(MUTABLE, DefaultLocation.EXCEPTION_PARAMETER);
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/igj/IGJAnnotatedTypeFactory.java]: Why imperatively here instead of declaratively with the annotation Mutable?

File: checker/src/org/checkerframework/checker/javari/JavariAnnotatedTypeFactory.java [hunk 3]
```diff
@@ -120,6 +122,14 @@ public JavariAnnotatedTypeFactory(BaseTypeChecker checker) {
     }
 
     @Override
+    protected void addCheckedCodeDefaults(QualifierDefaults defs) {
+        // An error is issued if the type of a thrown exception
+        // is not @Mutable.
+        defs.addCheckedCodeDefault(MUTABLE, DefaultLocation.EXCEPTION_PARAMETER);
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/javari/JavariAnnotatedTypeFactory.java]: Can the imperative code be replaced with a declarative annotation?

File: checker/src/org/checkerframework/checker/nullness/NullnessAnnotatedTypeFactory.java [hunk 4]
```diff
@@ -166,6 +168,14 @@ public NullnessAnnotatedTypeFactory(BaseTypeChecker checker, boolean useFbc) {
     }
 
     @Override
+    protected void addCheckedCodeDefaults(QualifierDefaults defs) {
+        // Exception parameters are always nonnull, so use that as the default
+        // rather than the CLIMB-to-the-top default.
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/nullness/NullnessAnnotatedTypeFactory.java]: Exception parameters are not handled by CLIMB-to-top: there is no expression that would refine that type.
Why is this here instead of as a declarative annotation?

File: framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java [hunk 5]
```diff
@@ -125,6 +125,9 @@
     /** Should use flow analysis? */
     protected boolean useFlow;
 
+    /** Is dataflow ever used for this checker? **/
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: I assume the following: we disable useFlow in certain situations and everUseDataFlow is the initial value of useFlow. Is that right?
Should the field be more symmetric to useFlow? Or is "DataFlow" something else?
The temporal dependency isn't clear from the name, but that is more of a problem with the lack of description of when "useFlow" is changed.

File: framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java [hunk 6]
```diff
@@ -400,6 +404,11 @@ protected void addCheckedCodeDefaults(QualifierDefaults defs) {
                                         + "@DefaultQualifierInHierarchy or @DefaultFor(DefaultLocation.OTHERWISE) not found. "
                                         + "Every checker must specify a default qualifier.");
         }
+        Set<? extends AnnotationMirror> tops = this.qualHierarchy.getTopAnnotations();
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: As tops and bottoms is only used in the condition, I would move their initialization there and limit their scope.

File: framework/src/org/checkerframework/framework/util/defaults/QualifierDefaults.java [hunk 7]
```diff
@@ -182,6 +192,31 @@ public void addUncheckedStandardDefaults(Iterable<? extends AnnotationMirror> to
...
+            for (AnnotationMirror top : tops) {
+                if (!conflictsWithExistingDefaults(checkedCodeDefaults, top, loc)) {
+                    addCheckedCodeDefault(top, loc);
+                }
+            }
+        }
+
+        for (DefaultLocation loc : standardClimbDefaultsBottom) {
+            for (AnnotationMirror bottom : bottoms) {
+                // Only add standard defaults in locations where a default has not be specified
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/defaults/QualifierDefaults.java]: I would remove the comment or move it to the first time this logic is used earlier in the method.
"be" -> "been"

File: framework/src/org/checkerframework/framework/util/defaults/QualifierDefaults.java [hunk 8]
```diff
@@ -105,6 +105,16 @@
     private final Map<Element, Boolean> elementAnnotatedFors = new IdentityHashMap<>();
 
     /**
+     * CLIMB-to-the-top locations whose standard default is top
+     */
+    public static final DefaultLocation[] standardClimbDefaultsTop = { DefaultLocation.LOCAL_VARIABLE, DefaultLocation.RESOURCE_VARIABLE,
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/defaults/QualifierDefaults.java]: So I had incorrectly assumed that "CLIMB-to-the-top" are only the locations that are flow-sensitively refined and forgot that "IMB" is for implicit bounds.
Local variables and resource variables are refined and therefore top.
Implicit bounds take the top (upper) and bottom (lower) qualifiers.
In most type systems exception parameters are top, unless the JLS gives us some property of the parameter, e.g. that the object is non-null.
I'm not sure "CLIMB-to-the-top" summarizes this well.

File: checker/src/org/checkerframework/checker/fenum/FenumAnnotatedTypeFactory.java (hunk 1 above)
Comment (smillst) [inline checker/src/org/checkerframework/checker/fenum/FenumAnnotatedTypeFactory.java]: @FenumUnqualified is annotated with @DefaultQualifierInHeirarchy, so it cannot also have @DefaultFor(EXCEPTION_PARAMETER).  This code is required in order to override the CLIMB-to-the-top defaults.  If you like this design, I'll add a comment.  

Alternatively, we could allow  @DefaultQualifierInHeirarchy and  @DefaultFor on the same annotation.  Or force type systems to use every CLIMB default or explicitly state all defaults.

File: framework/src/org/checkerframework/framework/util/defaults/QualifierDefaults.java (hunk 8 above)
Comment (smillst) [inline framework/src/org/checkerframework/framework/util/defaults/QualifierDefaults.java]: Is this a comment about the name of this field or a more general comment about CLIMB-to-top?  I used CLIMB in the code because that is how the standard checked code defaults are described in the manual.

**Comment by smillst** [comment]
(Pushed code changes to run tests. I'm still working on updating \@DefaultQualfierInHierarchy's documentation)

File: framework/src/org/checkerframework/framework/util/defaults/QualifierDefaults.java [hunk 9]
```diff
@@ -182,6 +192,32 @@ public void addUncheckedStandardDefaults(Iterable<? extends AnnotationMirror> to
     }
 
     /**
+     * Add standard CLIMB defaults that do not conflict with previously added defaults.
+     *
+     * @param tops    AnnotationMirrors that are top
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/util/defaults/QualifierDefaults.java]: I preseme this is "top for this type system" (and likewise for bottom)?

File: checker/src/org/checkerframework/checker/fenum/FenumAnnotatedTypeFactory.java [hunk 10]
```diff
@@ -34,6 +35,7 @@ public FenumAnnotatedTypeFactory(BaseTypeChecker checker) {
         this.postInit();
     }
 
+
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/fenum/FenumAnnotatedTypeFactory.java]: The only change in this file is the added import and this empty line. Is either one necessary?

File: checker/src/org/checkerframework/checker/igj/IGJAnnotatedTypeFactory.java [hunk 11]
```diff
@@ -27,6 +28,7 @@
 import org.checkerframework.framework.util.AnnotatedTypes;
 import org.checkerframework.framework.util.GraphQualifierHierarchy;
 import org.checkerframework.framework.util.MultiGraphQualifierHierarchy.MultiGraphFactory;
+import org.checkerframework.framework.util.defaults.QualifierDefaults;
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/igj/IGJAnnotatedTypeFactory.java]: These two added imports seem unnecessary, as no code was added.
Does your IDE provide "unused imports" warnings? Then the IDE would warn you about such unnecessary code.

File: checker/src/org/checkerframework/checker/javari/JavariAnnotatedTypeFactory.java [hunk 12]
```diff
@@ -34,6 +35,7 @@
 import org.checkerframework.framework.util.AnnotatedTypes;
 import org.checkerframework.framework.util.GraphQualifierHierarchy;
 import org.checkerframework.framework.util.MultiGraphQualifierHierarchy;
+import org.checkerframework.framework.util.defaults.QualifierDefaults;
 import org.checkerframework.javacutil.AnnotationUtils;
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/javari/JavariAnnotatedTypeFactory.java]: Unnecessary imports?

File: checker/src/org/checkerframework/checker/nullness/NullnessAnnotatedTypeFactory.java [hunk 13]
```diff
@@ -28,6 +29,7 @@
 import org.checkerframework.framework.type.typeannotator.TypeAnnotator;
 import org.checkerframework.framework.util.DependentTypes;
 import org.checkerframework.framework.util.MultiGraphQualifierHierarchy.MultiGraphFactory;
+import org.checkerframework.framework.util.defaults.QualifierDefaults;
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/nullness/NullnessAnnotatedTypeFactory.java]: Unnecessary imports.

File: checker/src/org/checkerframework/checker/nullness/qual/Nullable.java [hunk 14]
```diff
@@ -29,8 +29,6 @@
  */
 @SubtypeOf({})
 @ImplicitFor(trees = { Tree.Kind.NULL_LITERAL }, typeNames = { java.lang.Void.class })
-@DefaultFor({ DefaultLocation.LOCAL_VARIABLE, DefaultLocation.RESOURCE_VARIABLE,
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/nullness/qual/Nullable.java]: Now that these annotations are removed, the corresponding imports should also be removed.

File: checker/src/org/checkerframework/checker/oigj/ImmutabilityAnnotatedTypeFactory.java [hunk 15]
```diff
@@ -4,6 +4,7 @@
 import org.checkerframework.checker.oigj.qual.I;
 import org.checkerframework.checker.oigj.qual.Immutable;
 import org.checkerframework.checker.oigj.qual.Mutable;
+import org.checkerframework.checker.oigj.qual.OIGJMutabilityBottom;
 import org.checkerframework.checker.oigj.qual.ReadOnly;
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/oigj/ImmutabilityAnnotatedTypeFactory.java]: Unnecessary import!

File: checker/src/org/checkerframework/checker/oigj/OwnershipAnnotatedTypeFactory.java [hunk 16]
```diff
@@ -3,6 +3,7 @@
 import org.checkerframework.checker.oigj.qual.Dominator;
 import org.checkerframework.checker.oigj.qual.Modifier;
 import org.checkerframework.checker.oigj.qual.O;
+import org.checkerframework.checker.oigj.qual.OIGJMutabilityBottom;
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/oigj/OwnershipAnnotatedTypeFactory.java]: Unnecessary import!!

**Comment by wmdietl** [comment]
A few cosmetic things to cleanup. Otherwise, looks good to me.

**Comment by mernst** [comment]
Since there will be a number of commits that do and undo whitespace and import statements, I suggest that when this is ready to merge, you rebase (or use some other mechanism) to create a single commit that will result in a cleaner version control history.

File: checker/src/org/checkerframework/checker/oigj/ImmutabilityAnnotatedTypeFactory.java (hunk 15 above)
Comment (smillst) [inline checker/src/org/checkerframework/checker/oigj/ImmutabilityAnnotatedTypeFactory.java]: This one is actually need because I moved OIGJMutabilityBottom to the qual package.

File: checker/src/org/checkerframework/checker/oigj/OwnershipAnnotatedTypeFactory.java (hunk 16 above)
Comment (smillst) [inline checker/src/org/checkerframework/checker/oigj/OwnershipAnnotatedTypeFactory.java]: This one is actually need because I moved OIGJMutabilityBottom to the qual package.

**Comment by smillst** [comment]
I rebased these changes and created a new pull request #567.  I'll merge the new pull request once Travis passes.
